Store/StoreCart.php

```diff
@@ -14,736 +14,638 @@
+        $valid = true;
+
+        if ($entry->quantity <= 0) {
+            $message = sprintf(
+                Store::_('Quantity of “%s” item #%s must be at least one.'),
+                $entry->item->product->title,
+                $entry->item->sku
+            );
+
+            $this->addMessage(new SwatMessage($message, SwatMessage::ERROR));
+            $valid = false;
+        }
+
+        return $valid;
+    }
+
+    /**
+     * Checks to see if an entry is valid after combining the entry with
+     * existing entries in this cart.
+     *
+     * Used to verify quantities meet requirments for purchase.
+     *
+     * While validating an entry, this method should add validation error
+     * messages to this cart. See {@link StoreCart::addMessage()}.
+     *
+     * @param StoreCartEntry $entry the cart entry to validate when combined
+     *                              with the other entries in this cart
+     *
+     * @return bool true if the entry is valid in combination with the
+     *              entries in this cart and false if it is not valid
+     */
+    protected function validateCombinedEntry(StoreCartEntry $entry)
+    {
+        return true;
+    }
+
+    /**
+     * Performs pre-save processing on a single entry.
+     *
+     * This is meant to be called on all entries before they are saved.
+     *
+     * @param StoreCartEntry $entry the entry to process
+     */
+    protected function preSaveEntry(StoreCartEntry $entry)
+    {
+        $entry->instance = $this->app->getInstance();
+    }
+
+    /**
+     * Adds a new entry to this cart.
+     *
+     * This performs maintenance tasks required when adding an actual new
+     * entry object to this cart.
+     *
+     * @param StoreCartEntry $entry the entry to add
+     */
+    protected function addNewEntry(StoreCartEntry $entry)
+    {
+        $this->preSaveEntry($entry);
+        $entry->save();
+
+        $this->entries[] = $entry;
+        $this->entries_by_id[$entry->id] = $entry;
+
+        if ($this->module instanceof StoreCartModule) {
+            $this->module->registerAddedEntry($entry);
+        }
+
+        $this->setChanged();
+    }
+
+    /**
+     * Sorts the entries of this cart.
+     *
+     * This method is called by getEntries() if the cart's sort order is stale.
+     */
+    protected function sort()
+    {
+        if (!$this->sorted) {
+            $max_ids = [];
+            foreach ($this->entries as $entry) {
+                $key = $entry->item->product->id;
+
+                if (!isset($max_ids[$key])) {
+                    $max_ids[$key] = $entry->id;
+                } else {
+                    $max_ids[$key] = max($entry->id, $max_ids[$key]);
+                }
+            }
+
+            foreach ($this->entries as $entry) {
+                $key = $entry->item->product->id;
+                $entry->setProductMaxCartEntryId($max_ids[$key]);
+            }
+
+            usort($this->entries, ['StoreCart', 'compare']);
```

Comment:
Should this be StoreCart::class?

Reply:
I think it's static, but maybe we can still do the same. Let's leave it for another cleanup PR.